jni/lib/src/jni_object.dart

```diff
@@ -0,0 +1,175 @@
+import 'dart:ffi';
+
+import 'package:ffi/ffi.dart';
+
+import 'third_party/jni_bindings_generated.dart';
+import 'extensions.dart';
+import 'jni_class.dart';
+import 'jvalues.dart';
+import 'jni_exceptions.dart';
+
+part 'jni_object_methods_generated.dart';
+
+/// JniObject is a convenience wrapper around a JNI local object reference.
+///
+/// It holds the object, its associated associated jniEnv etc..
+/// It should be distroyed with [delete] method after done.
+///
+/// It's valid only in the thread it was created.
+/// When passing to code that might run in a different thread (eg: a callback),
+/// consider obtaining a global reference and reconstructing the object.
+class JniObject {
+  JClass _cls;
+  final JObject _obj;
+  final Pointer<JniEnv> _env;
+  bool _deleted = false;
+  JniObject.of(this._env, this._obj, this._cls);
+
+  @pragma('vm:prefer-inline')
+  void _checkDeleted() {
+    if (_deleted) {
+      throw UseAfterFreeException(this, _obj);
+    }
+  }
+
+  JniObject.fromJObject(Pointer<JniEnv> env, JObject obj)
+      : _env = env,
+        _obj = obj,
+        _cls = nullptr;
+
+  /// Reconstructs a JniObject from [r]
+  ///
+  /// [r] still needs to be explicitly deleted when
+  /// it's no longer needed to construct any JniObjects.
+  JniObject.fromGlobalRef(Pointer<JniEnv> env, JniGlobalObjectRef r)
+      : _env = env,
+        _obj = env.NewLocalRef(r._obj),
+        _cls = env.NewLocalRef(r._cls) {
+    if (r._deleted) {
+      throw UseAfterFreeException(r, r._obj);
+    }
+  }
+
+  /// Delete the local reference contained by this object.
+  ///
+  /// Do not use a JniObject after calling [delete].
+  void delete() {
+    if (_deleted == true) {
+      throw DoubleFreeException(this, _obj);
+    }
+    _env.DeleteLocalRef(_obj);
+    if (_cls != nullptr) {
+      _env.DeleteLocalRef(_cls);
+    }
+    _deleted = true;
+  }
+
+  /// Returns underlying [JObject] of this [JniObject].
+  JObject get jobject {
+    _checkDeleted();
+    return _obj;
+  }
+
+  /// Returns underlying [JClass] of this [JniObject].
+  JObject get jclass {
+    _checkDeleted();
+    if (_cls == nullptr) {
+      _cls = _env.GetObjectClass(_obj);
+    }
+    return _cls;
+  }
+
+  /// Get a JniClass of this object's class.
+  JniClass getClass() {
+    _checkDeleted();
+    if (_cls == nullptr) {
+      return JniClass.of(_env, _env.GetObjectClass(_obj));
```

Comment:
I am not saving it in original object, so I think this would be fine.

I need to do NewLocalRef in second case, because `JniObject._cls` also gets deleted (`DeleteLocalRef`) on `JniObject.delete`.